File: leetcode/leetcode_py/zigzag-conversion.py
'''
https://leetcode.com/problems/zigzag-conversion

Solution. Traverse and store on level
time: o(n)
space: o(n)
'''
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1: return s
        dir = True
        level = 0
        slot = [""] * numRows
        for c in s:
            slot[level] += c
            level += 1 if dir else -1
            if level == numRows-1 or level == 0:
                dir = not dir

        res = ""
        for values in slot:
            res += values
        return res
        

File: leetcode/leetcode_cpp/zigzag-conversion.cpp
/*
https://leetcode.com/problems/zigzag-conversion

Solution. Traverse and store on level
time: o(n)
space: o(n)
*/
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1) return s;
        bool dir = true;
        int level = 0;
        vector<string> slot(min(numRows, int(s.size())));
        for (auto c : s) {
            slot[level] += c;
            level += dir ? 1 : -1;
            if (level == numRows-1 || level == 0) dir = !dir;
        }

        string res;
        for (auto values : slot)
            res += values;
        return res;
    }
};